### PR TITLE
Fixes and improvements for Run Code Analysis command

### LINKED ISSUE #26817: Analyzer warnings disappear from Error List when the file with the warning is opened
**Version Used**: VS 15.8.0 Preview 1.1, .Net Core SDK 2.1.4

**Steps to Reproduce**:

1. `dotnet new classlib`
2. `dotnet add package Microsoft.CodeQuality.Analyzers`
3. Change code in Class1.cs to:

    ```c#
    namespace Foo
    {
        public interface IFoo { }
    }
    ```
4. Open the csproj in VS.
5. Build Solution.
6. Open Class1.cs.
7. Wait a few seconds.

**Expected Behavior**: Whether the analyzer warning "CA1040 Avoid empty interfaces" appears in the Error List should not depend on whether the relevant file is open.

**Actual Behavior**: When I Build Solution, the analyzer warning "CA1040 Avoid empty interfaces" appears in the Error List. When I open the relevant file, the warning disappears from the Error List after few seconds.

This only happens in the Build + IntelliSense mode. In the Build Only mode, the warning is always visible

---

I'm assuming this is a general issue. If it's actually an issue with the specific analyzer, I can move this issue to [the dotnet/roslyn-analyzers repo](https://github.com/dotnet/roslyn-analyzers).

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #26817
Fixes [AB#1698190](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1698190)

This PR fixes the regression where in diagnostics reported for a project/solution after executing `Run Code Analysis` command were getting cleared out after editing source files and/or switching active file tabs.

While fixing the above, I also enhanced the `Run Code Analysis` to handle multi-tfm projects so that now we run analysis for all tfm flavors of a multi-tfm project instead of just the current tfm.

### CODE REVIEW

**Comment by equin0x80** [comment]
Thank you. A lot of people wanted this one!
Will this be released in the VS2022 update stream or is it going to be mysteriously held back for a future VS version? (And in general for the future: how can I know this, by looking at an issue report?)

**Comment by CyrusNajmabadi** [comment]
This is being merged into main, which flows into the next vs release.

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs [hunk 1]
```diff
@@ -48,14 +53,12 @@ internal partial class DiagnosticIncrementalAnalyzer
                     return existingData;
                 }
 
-                // Perf optimization: Check whether analyzer is suppressed for project or document and avoid getting diagnostics if suppressed.
-                if (!DocumentAnalysisExecutor.IsAnalyzerEnabledForProject(stateSet.Analyzer, document.Project, GlobalOptions) ||
+                // Check whether analyzer is suppressed for project or document and avoid getting diagnostics if suppressed.
+                // Note that we do not return empty DocumentAnalysisData for suppressed analyzer as that would clear the
+                // error list and remove the previously reported diagnostics from "Run Code Analysis" command.
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs:58]: i think this is a bit too vague here.  I have no idea in the context of this code what the "run code analysis" command is or why it has this special capability.  What *does* end up clearing the results of it?  something feels *very* stateful here wrt to the error list and thsee features, adn i don't have a clear sense of how that state is managed and what is or isn't ok.

Note: this also seems like a scenario that i have no idea how it will work with pull diags.

**Comment by CyrusNajmabadi** [comment]
Probably best to have a teams chat over this to get a better sense of what's going on.  Also should talk with David to understand how this needs to work in the LSP pull future.  "run code analysis" is def somethign i'm not familiar with, so i'm not sure if/how it's intended to work today or in the future.

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs (hunk 1 above)
Comment (mavasani) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs:58]: > I have no idea in the context of this code what the "run code analysis" command is or why it has this special capability. What does end up clearing the results of it?

I'll end up adding a better comment here. "Run code analysis" is just one of the scenarios, but the core issue was prior code was essentially stating - there are no diagnostics in the document that we are trying to fetch cached diagnostics for, which was not true. This ends up pushing empty diagnostics for that document to the error list. The new code is basically returning null (indicating there are no cached diagnostics) and setting a flag that clearly indicates that analysis is disabled for this document, so the caller is free to perform any optimizations around analysis of that document, such as skipping running analysis for it.

Comment (mavasani) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs:58]: > Note: this also seems like a scenario that i have no idea how it will work with pull diags.

Tracked by https://github.com/dotnet/roslyn/issues/65967. However, given the large number of customer reports about this regression, we ought to fix this asap for push diagnostics, we cannot wait until pull diagnostics are enabled and #65967 is fixed.

**Comment by mavasani** [comment]
> Probably best to have a teams chat over this to get a better sense of what's going on. Also should talk with David to understand how this needs to work in the LSP pull future. "run code analysis" is def somethign i'm not familiar with, so i'm not sure if/how it's intended to work today or in the future.

Sure, makes sense.

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs (hunk 1 above)
Comment (equin0x80) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs:58]: > i think this is a bit too vague here. I have no idea in the context of this code what the "run code analysis" command is or why it has this special capability. What _does_ end up clearing the results of it? something feels _very_ stateful here wrt to the error list and thsee features, adn i don't have a clear sense of how that state is managed and what is or isn't ok.
> 
> Note: this also seems like a scenario that i have no idea how it will work with pull diags.

As a small-scale self-employed nobody who actually uses your VS Community tool and doesn't work for Microsoft, or have friends at MS, but *does* use VS for 80% of the programming work that I do in my career -- let me explain the issue. Listen to your user. (You used to call us "Elvises", remember that?)

My assumption (and forgive me if it's wrong, but explain why --) is that you added this code analysis feature for Enterprise users, and we get it for free in Community (which is amazing, I love VS) but it wasn't properly tested there.

Here is how I can *repeatably* reproduce the annoyance. I open my recent VS (it is 2022 64-bit, 17.4.3, with .NET Framework 4.8.04084). I load my primary work project, it's nothing magic, it's just a nice large Windows Forms application that I began in 2016. And I do: Analyze (menu) > Code Analysis > On Solution. Then I am shown the useful output. But once I double-click any entry and start to look at a single source file, then all the other analysis (other files) disappears.

Thanks for listening to your user.

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs:58]: @equin0x80 my messages were about lack of clarity on what this fix was doing. Thanks!

Comment (mavasani) [reply src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs]: I have clarified the comment and also have a descriptive doc comment on the method.

File: src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs [hunk 2]
```diff
@@ -34,10 +36,13 @@ internal partial class DiagnosticIncrementalAnalyzer
             CompilerDiagnosticsScope compilerDiagnosticsScope,
             bool isActiveDocument, bool isVisibleDocument,
             bool isOpenDocument, bool isGeneratedRazorDocument,
-            CancellationToken cancellationToken)
+            CancellationToken cancellationToken,
+            out bool skipAnalysisForNonCachedDocument)
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs:40]: nit - maybe a more clear name would be isAnalyzerEnabled / isAnalyzerSuppressed?

The current name is a little hard to parse (skip...noncached)

**Review by dibarbet — APPROVED** [review]
